src/box/lua/config/utils/schema.lua

```diff
@@ -1389,6 +1389,133 @@ end
+    -- context of descendant schema nodes. Don't track them.
+    local ignored_annotations = {
+        allowed_values = true,
+        validate = true,
+        default = true,
+        apply_default_if = true,
+    }
+
+    local frame = table.copy(ctx.annotation_stack[#ctx.annotation_stack] or {})
+    for k, v in pairs(schema) do
+        if not non_annotation_keys[k] and not ignored_annotations[k] then
+            frame[k] = v
+        end
+    end
+    table.insert(ctx.annotation_stack, frame)
+end
+
+-- Step up from a schema node.
+--
+-- Returns the computed fields for the node that we're leaving.
+--
+-- See preprocess_enter() for details.
+local function preprocess_leave(ctx)
+    return {
+        annotations = table.remove(ctx.annotation_stack),
+    }
+end
+
+-- The function prepares the given schema node tree in the
+-- following way.
+--
+-- * The schema node is copied.
+-- * All its descendant nodes are prepared using this algorithm
+--   (recursively).
+-- * Computed fields are calculated and stored in the copied node.
+-- * The copied node is the result.
+--
+-- The sketchy structure of the returned schema node is the
+-- following.
+--
+-- {
+--     <..fields..>
+--     computed = {
+--         annotations = <...>,
+--     },
+-- }
+--
+-- At now there is only one computed field: `annotations`.
+--
+-- The `annotations` field contains all the annotations merged
+-- from the root schema node down to the given one. If the same
+-- annotation is present in an ancestor node and in an descendant
+-- node, the latter is preferred.
+--
+-- Design details
+-- --------------
+--
+-- The copying is performed to decouple schema node definitions
+-- provided by a caller from the schema nodes stored in the schema
+-- object. It allows to modify them without modifying caller's own
+-- objects.
+--
+-- It is especially important if parts of one schema are passed
+-- as schema node definitions to create another schema.
+--
+-- This case also motivates to store all the computed fields in
+-- the `computed` field. This way it is easier to guarantee that
+-- all the computed fields are stripped down from the original
+-- schema nodes and don't affect the new schema anyhow.
+local function preprocess_schema(schema, ctx)
+    local res = table.copy(schema)
+
+    res.computed = nil
+    preprocess_enter(ctx, res)
+
+    -- luacheck: ignore 542 empty if branch
+    if is_scalar(schema) then
+        -- Nothing to do.
+    elseif schema.type == 'record' then
+        local fields = {}
+        for field_name, field_def in pairs(schema.fields) do
+            fields[field_name] = preprocess_schema(field_def, ctx)
+        end
+        res.fields = fields
+    elseif schema.type == 'map' then
+        res.key = preprocess_schema(schema.key, ctx)
+        res.value = preprocess_schema(schema.value, ctx)
+    elseif schema.type == 'array' then
+        res.items = preprocess_schema(schema.items, ctx)
+    else
+        assert(false)
```

Comment:
I think it looks better to replace this `assert(false)` with with `assert(is_scalar(schema))` and remove the corresponding 'if'.

Reply:
I have structured all the walkthroughs this way: scalar first, then record, map, array. The same empty if branch is exist in `filter_impl()` for example. I hope that a repeated structure of the code makes it easier to read at whole.